### PR TITLE
[ABM-Edge-3] Implement the payment service

### PR DESCRIPTION
**Pull request by Shabirmean**
### Depends on:
- Please review _(or make sure **merged** status of)_ the following PR before starting on this: 
   - #2  

### What's this?
- Implementation of the `payment` service APIs
- Some minor updates to the inventory service

### Architecture
image

***Notes on the Architecture:***
- The initial MVP will implement the data-store in memory. We will work on the expanding the app with an external storage option after the MVP
- The app exposes a _/switch_ functionality to switch between store types
- More details at: [go/abm-edge-app](http://go/abm-edge-app)

### CODE REVIEW

**Review by bourgeoisor — COMMENTED** [review]
Looks good! Couple comments.

Additional comment: highly recommend adding class and method comments! If not for yourself, for your future self and all others reading this code someday.

File: point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/PaymentsController.java [hunk 1]
```diff
@@ -50,8 +78,21 @@ public String readiness() {
...
   }
 
-  @RequestMapping("/")
-  public String home() {
-    return "Hello Anthos BareMetal - Payments Controller";
+  @PostMapping(value = "/pay")
+  public ResponseEntity<String> pay(@RequestBody Payment payment) {
+    Bill bill = this.activePaymentGateway.pay(payment);
+    String jsonString = GSON.toJson(bill, Bill.class);
+    return new ResponseEntity<>(jsonString, HttpStatus.OK);
```
[the remark this answers is not in the document]
Comment (Shabirmean) [reply point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/PaymentsController.java:99]: - If the payload doesn't match:
      - There are a bunch of handlers setup by Springboot before the request reaches this point. So if the incoming request cannot be deserialized into a `Payment` object then earlier handlers in the request path will fail and a non `HTTP-OK` error will be returned 

- `activePaymentGateway.pay()` and `GSON.toJson()`
     - I have added a checked exception for these cases in https://github.com/GoogleCloudPlatform/anthos-edge-usecases/pull/3/commits/a7590ffbcafb8673f842b14d9c8548c7c0f69d32 and https://github.com/GoogleCloudPlatform/anthos-edge-usecases/pull/3/commits/f1d947d1074dabcd266ced833f67ddbcbcd29acd

File: point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/dao/InMemoryPaymentGateway.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  public Bill pay(Payment payment) {
+    UUID pId = payment.getId();
+    paymentMap.put(pId, payment);
+    Pair<String, String> generatedBill = generateBill(pId, payment);
+    Bill paymentBill = new Bill();
+    paymentBill.setPayment(payment);
+    paymentBill.setStatus(PaymentStatus.SUCCESS);
+    paymentBill.setBalance(Double.valueOf(generatedBill.getRight()));
+    paymentBill.setPrintedBill(generatedBill.getLeft());
+    return paymentBill;
```
[the remark this answers is not in the document]
Comment (Shabirmean) [reply point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/dao/InMemoryPaymentGateway.java]: Nice. Ive changed it like this in: c2be7565f4ebf891307cd893d13696c1be6ed0af

File: point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/dao/PaymentGateway.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.abmedge.payments.dao;
+
+import com.google.abmedge.payments.dto.Bill;
+import com.google.abmedge.payments.dto.Payment;
+
+public interface PaymentGateway {
+
```
[the remark this answers is not in the document]
Comment (Shabirmean) [reply point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/dao/PaymentGateway.java:22]: I was just using the code formatter that comes by default on the IDE. But this is awesome. I have set it up and fixed some issues in: 50d64d2249d037215b9f5fd126c80f119ded595c

File: point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/dao/InMemoryPaymentGateway.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  private Pair<String, String> generateBill(UUID pId, Payment payment) {
```
[the remark this answers is not in the document]
Comment (Shabirmean) [reply point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/dao/InMemoryPaymentGateway.java]: Yup. I have completely refactored this code to be more readable in: 9341853aa54f8e53cf3d29df4f546a16a37d389c. See it context [here](https://github.com/GoogleCloudPlatform/anthos-edge-usecases/blob/add-payment-apis/point-of-sale-app/payments/src/main/java/com/google/abmedge/payments/dao/InMemoryPaymentGateway.java#L51). Let me know if this is good!

**Review by bourgeoisor — DISMISSED** [review]
Looks a lot better, thank you for the comments + linting!

**Review by puneith — APPROVED** [review]

**Review by bourgeoisor — APPROVED** [review]
LGTM!

**Pull request merged**